esi/lib/nodes.py

```diff
@@ -16,6 +16,48 @@
 from esi.lib import networks
 
 
+def node_list_full_info(connection):
+    """Get a list of nodes with more attributes
+
+    :param connection: An ESI connection
+    :type connection: :class:`~esi.connection.ESIConnection`
+
+    :returns: A list of dictionaries containing baremetal and lease information
+    """
+
+    esi_nodes = []
+    baremetal_nodes = []
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        f1 = executor.submit(connection.list_nodes)
+        f2 = executor.submit(connection.baremetal.nodes, details=True)
+        esi_nodes = f1.result()
+        baremetal_nodes = list(f2.result())
+
+    esi_nodes.sort(key=lambda node: node.name)
+    baremetal_nodes.sort(key=lambda node: node.name)
```

Comment:
I admit that this sort of implicit matching makes me nervous, especially because I think Ironic node names don't actually have to be unique. In addition it depends on the two services returning the same list of nodes - which they *should* do, but if the behavior changes in one then the results will be messed up.